pkg/agent/controller/egress/ipassigner/ip_assigner_linux.go

```diff
@@ -34,153 +30,146 @@ import (
+	link, err := netlink.LinkByName(deviceName)
+	if err == nil {
+		return link, nil
 	}
-	a.mutex.Lock()
-	defer a.mutex.Unlock()
-	for _, file := range files {
-		egressName := file.Name()
-		fileName := ipSavedFile(a.egressRunDir, egressName)
-		if ip, err := ioutil.ReadFile(fileName); err == nil {
-			a.assignedEgress[egressName] = string(ip)
-		}
+	dummy := &netlink.Dummy{
+		LinkAttrs: netlink.LinkAttrs{Name: deviceName},
 	}
-	return
+	if err = netlink.LinkAdd(dummy); err != nil {
+		return nil, err
+	}
+	return dummy, nil
 }
 
-// AssignEgressIP assign an egressIP and save persistent files.
-func (a *ipAssigner) AssignEgressIP(egressIP, egressName string) error {
-	egressSpecIP := net.ParseIP(egressIP)
-	if egressSpecIP == nil {
-		return fmt.Errorf("invalid IP %s", egressIP)
-	}
-	addr := netlink.Addr{IPNet: &net.IPNet{IP: egressSpecIP, Mask: net.CIDRMask(32, 32)}}
-	ifaceName := a.egressInterface.Name
-	if err := netlink.AddrAdd(a.egressLink, &addr); err != nil {
-		return fmt.Errorf("failed to add ip %v to interface %s: %v", addr, ifaceName, err)
-	}
-	if err := a.saveIPAssignFile(egressIP, egressName); err != nil {
-		return fmt.Errorf("failed to save Egress IP assign, Egress %s, IP: %s: %v", egressName, egressIP, err)
+// loadIPAddresses gets the IP addresses on the dummy device and caches them in memory.
+func (a *ipAssigner) loadIPAddresses() error {
+	addresses, err := netlink.AddrList(a.dummyDevice, netlink.FAMILY_ALL)
+	if err != nil {
+		return err
 	}
-	isIPv4 := egressSpecIP.To4()
-	if isIPv4 != nil {
-		if err := arping.GratuitousARPOverIface(isIPv4, a.egressInterface); err != nil {
-			return fmt.Errorf("failed to send gratuitous ARP: %v", err)
-		}
-		klog.V(2).InfoS("Sent gratuitous ARP", "ip", isIPv4)
-	} else if isIPv6 := egressSpecIP.To16(); isIPv6 != nil {
-		return fmt.Errorf("failed to send Advertisement: %v", ipv6NotSupportErr)
+	newAssignIPs := sets.NewString()
+	for _, address := range addresses {
+		newAssignIPs.Insert(address.IP.String())
 	}
-	klog.V(2).InfoS("Added ip", "ip", addr, "interface", ifaceName)
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	a.assignedIPs = newAssignIPs
 	return nil
 }
 
-// UnassignEgressIP unassign an egressIP and delete the persistent files.
-func (a *ipAssigner) UnassignEgressIP(egressName string) error {
-	egressIP, has := a.isAssignedIP(egressName)
-	if !has {
-		return nil
+// AssignIP ensures the provided IP is assigned to the dummy device.
+func (a *ipAssigner) AssignIP(ip string) error {
+	parsedIP := net.ParseIP(ip)
+	if parsedIP == nil {
+		return fmt.Errorf("invalid IP %s", ip)
 	}
-	egressSpecIP := net.ParseIP(egressIP)
-	addr := netlink.Addr{IPNet: &net.IPNet{IP: egressSpecIP, Mask: net.CIDRMask(32, 32)}}
-	ifaceName := a.egressLink.Attrs().Name
-	if err := netlink.AddrDel(a.egressLink, &addr); err != nil {
-		return fmt.Errorf("failed to delete ip %v from interface %s: %v", addr, ifaceName, err)
+
+	if err := func() error {
```

Comment:
new version LGTM. The use of the RW mutex makes more sense now too, given that we keep the write lock for an expensive operation now (call to `netlink.AddrAdd`).